File: 118-PascalsTriangle/118-PascalsTriangle.java
/*

118. Pascal's Triangle
Easy

Given a non-negative integer numRows, generate the first numRows of Pascal's triangle.

In Pascal's triangle, each number is the sum of the two numbers directly above it.

Example:

Input: 5
Output:
[
     [1],
    [1,1],
   [1,2,1],
  [1,3,3,1],
 [1,4,6,4,1]
]

*/
public class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> resList = new ArrayList<List<Integer>>();
        if (numRows <= 0) return resList;
        for (int i = 0; i < numRows; i++){
            List<Integer> row =  new ArrayList<Integer>();
            for (int j = 0; j < i+1; j++){
                if (j==0 || j==i) row.add(1);
                else row.add(resList.get(i-1).get(j-1) + resList.get(i-1).get(j));
            }
            resList.add(row);
        }
        return resList;
    }
}

File: 118-PascalsTriangle/118-PascalsTriangle.py
"""

118. Pascal's Triangle

Given an integer numRows, return the first numRows of Pascal's triangle.

In Pascal's triangle, each number is the sum of the two numbers directly above it as shown:

Example 1:

Input: numRows = 5
Output: [[1],[1,1],[1,2,1],[1,3,3,1],[1,4,6,4,1]]

Example 2:

Input: numRows = 1
Output: [[1]]

Constraints:

    1 <= numRows <= 30

"""

class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        if numRows <= 1: return [[1]]
        res = [[1], [1, 1]]
        if numRows == 2: return res
        for i in range(2, numRows):
            temp = []
            for j in range(i + 1):
                if j == 0 or j == i: temp.append(1)
                else: temp.append(res[i - 1][j - 1] + res[i - 1][j])
            res.append(temp)
        return res
